### PR TITLE
[WIP] Use AOT compiled builds.

### PR DESCRIPTION
**Pull request by alloy**
## Why

* Clear mapping between the _actual_ code that gets run (JS) and code on-disk. (This is what’s currently most helpful to me, as it makes debugging and reasoning about code simpler.)
* Faster startup time.
* No more lazy compilation, akin to warming up a cache.
* No in-memory compiled JS, which can take quite some memory.

  Difference in post-startup memory consumption:
  - Before: ~230MB
  - After: ~100MB

  _Note that because of lazy compilation, it could well be that more source is loaded **during** runtime, which would make this difference even larger. At the moment it is not clear what portion of the server-side code is loaded at startup time and what code is loaded at later times._

## TODOs

- [x] Compile CoffeeScript files.
- [x] Compile ES6 JS files.
- [x] Remove file extensions from `require` statements.
- [x] Remove full paths to packages in `node_modules`.
- [x] Sync `public` assets.
- [ ] Compile Jade files, currently these are copied over as-is. I haven’t delved into Jade enough yet to know how the system works exactly, but if I’m not mistaken there’s JIT compilation of these templates occurring during server-side renders, these can probably be AOT compiled too? Afaik we currently only AOT compile browser assets.
- [ ] There’s some issues with loading some precompiled CoffeeScript. E.g. `desktop/lib/template_modules.coffee` ends up in the `desktop/assets/main_layout.js` Browserify bundle in such a way that it tries to store required modules on `this` but `this` is `undefined`.
- [ ] Some of our own npm packages we use also contain uncompiled CoffeeScript, these also need to be updated.
- [ ] Probably some unknowns…

### CODE REVIEW

File: desktop/apps/about/client/index.coffee [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Backbone = require 'backbone'
-AboutRouter = require './router.coffee'
+AboutRouter = require './router'
```
Comment (craigspaeth) [inline desktop/apps/about/client/index.coffee]: Thanks! Removing the need for .coffee extensions is something I've wanted to do for a while. I assume this works in development for client-side code too? If not one can set it to be supported [in browserify](https://github.com/substack/node-browserify#browserifyfiles--opts).

**Comment by craigspaeth** [comment]
That's right wrt Jade templates—they are JIT compiled on the server and AOT on client. AOT compiling templates in Node isn't something I've heard of being very popular, but a quick googling finds [this module](https://github.com/axln/compiled-jade-render) that might be worth looking into.

That browserify issue is strange—maybe related to my comment above about extensions?

👏  very exciting. Hopefully this helps clean up some noise in memory debugging. FWIW it also occured to me that with our introduction of async/await and native promises in Force we might might want to try using [Bluebird](http://bluebirdjs.com/docs/getting-started.html) in place of the native promise object, e.g. `global.Promise = require('bluebird')`. The last big memory leak in Force was due to Q's poor resource performance, and I don't think native promises [are immune to that either](https://github.com/nodejs/node/issues/6673).

**Comment by alloy** [comment]
> a quick googling finds this module that might be worth looking into.

Nice find! Seems to be a packaged version of the description here http://stackoverflow.com/a/37821914/95397

**Comment by alloy** [comment]
It’s looking likely that I’ll have to switch to using something like grunt. E.g. the `jade` CLI tool doesn’t support saving compiled source to a pathname that reflects the original pathname (only an immediate out dir). And it’s probably also going to take a lot longer to compile all Jade templates by invoking the CLI for each individual file vs a single cached compiler instance.

**Comment by craigspaeth** [comment]
I'm reluctant to introduce gulp/grunt as I'd prefer to keep our build tools simple, but we _do_ now have a [scripts](https://github.com/artsy/force/tree/master/scripts) directory and we could start expanding inside of with gulp or [vanilla js scripts](http://stackoverflow.com/questions/16509848/running-node-from-a-bash-script). I unassigned myself from this PR, but feel free to re-assign me when ready.

**Comment by alloy** [comment]
Ah yeah, I’m happy to use plain scripts over gulp 👍

I’ll close this PR while it’s on the back burner.

**Pull request closed without merging**